services/graph/pkg/service/v0/api_drives_drive_item.go

```diff
@@ -0,0 +1,312 @@
+type DrivesDriveItemService struct {
+	logger           log.Logger
+	gatewaySelector  pool.Selectable[gateway.GatewayAPIClient]
+	identityCache    identity.IdentityCache
+	resharingEnabled bool
+}
+
+// NewDrivesDriveItemService creates a new DrivesDriveItemService
+func NewDrivesDriveItemService(logger log.Logger, gatewaySelector pool.Selectable[gateway.GatewayAPIClient], identityCache identity.IdentityCache, resharing bool) (DrivesDriveItemService, error) {
+	return DrivesDriveItemService{
+		logger:           log.Logger{Logger: logger.With().Str("graph api", "DrivesDriveItemService").Logger()},
+		gatewaySelector:  gatewaySelector,
+		identityCache:    identityCache,
+		resharingEnabled: resharing,
+	}, nil
+}
+
+func (s DrivesDriveItemService) UnmountShare(ctx context.Context, resourceID storageprovider.ResourceId) error {
+	gatewayClient, err := s.gatewaySelector.Next()
+	if err != nil {
+		return err
+	}
+
+	// This is a a bit of a hack. We should not rely on a specific format of the item id.
+	// (But currently the ShareID is
+	shareId := resourceID.GetOpaqueId()
+
+	// Now, find out the resourceID of the shared resource
+	getReceivedShareResponse, err := gatewayClient.GetReceivedShare(ctx,
+		&collaboration.GetReceivedShareRequest{
+			Ref: &collaboration.ShareReference{
+				Spec: &collaboration.ShareReference_Id{
+					Id: &collaboration.ShareId{
+						OpaqueId: shareId,
+					},
+				},
+			},
+		},
+	)
+	if errCode := errorcode.FromCS3Status(getReceivedShareResponse.GetStatus(), err); errCode != nil {
+		s.logger.Debug().Err(errCode).
+			Str("shareid", shareId).
+			Msg("failed to read share")
+		return errCode
+	}
+
+	// Find all accepted shares for this resource
+	receivedSharesResponse, err := gatewayClient.ListReceivedShares(ctx, &collaboration.ListReceivedSharesRequest{
+		Filters: []*collaboration.Filter{
+			{
+				Type: collaboration.Filter_TYPE_STATE,
+				Term: &collaboration.Filter_State{
+					State: collaboration.ShareState_SHARE_STATE_ACCEPTED,
+				},
+			},
+			{
+				Type: collaboration.Filter_TYPE_RESOURCE_ID,
+				Term: &collaboration.Filter_ResourceId{
+					ResourceId: getReceivedShareResponse.GetShare().GetShare().GetResourceId(),
+				},
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	var errs []error
+
+	// Reject all the shares for this resource
+	for _, receivedShare := range receivedSharesResponse.GetShares() {
+		receivedShare.State = collaboration.ShareState_SHARE_STATE_REJECTED
+
+		updateReceivedShareRequest := &collaboration.UpdateReceivedShareRequest{
+			Share:      receivedShare,
+			UpdateMask: &fieldmaskpb.FieldMask{Paths: []string{_fieldMaskPathState}},
+		}
+
+		_, err := gatewayClient.UpdateReceivedShare(ctx, updateReceivedShareRequest)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+	}
+
+	// We call it a success if all shares could successfully be rejected, otherwise
+	// we return an error
+	if len(errs) != 0 {
+		return errors.Join(errs...)
```

Comment:
errors.Join(errs...) ignores nil errors and returns nil in that case, no err len check needed, just return it